hail/python/hail/expr/expressions/typed_expressions.py

```diff
@@ -3055,6 +2998,188 @@ def includes_end(self):
         return self._method("includesEnd", tbool)
 
 
+class NDArrayExpression(Expression):
+    """Expression of type :class:`.tndarray`.
+
+    >>> names = hl._ndarray([['Alice', 'Bob'], ['Carol', 'Dan']])
+    """
+
+    @property
+    def ndim(self):
+        """The number of dimensions of this ndarray.
+
+        Examples
+        --------
+
+        >>> names.ndim
+        2
+
+        Returns
+        -------
+        :obj:`int`
+        """
+        return self._type.ndim
+
+    @typecheck_method(item=oneof(expr_int64, tupleof(expr_int64)))
+    def __getitem__(self, item):
+        if not isinstance(item, tuple):
+            item = (item,)
+
+        if len(item) != self.ndim:
+            raise ValueError(f'Must specify one index per dimension. '
+                             f'Expected {self.ndim} dimensions but got {len(item)}')
+
+        # indexes must be iterable
+        if len(item) == 0:
+            idxs = hl.empty_array(hl.tint64)
+        else:
+            idxs = hl.array(list(item))
```

Comment:
I don't totally understand why the index is an array IR, instead of a sequence of integers. This also makes it very difficult to build slicing on top. Let's discuss tomorrow.